shim/shim.c

```diff
@@ -270,38 +270,68 @@ handle_signals(char *container_id, int outfd) {
 }
 
 /*!
- * Read data from infd and write to outfd
+ * Read data from stdin(with tty set in raw mode)
+ * and send it to proxy I/O channel
+ * Reference : https://github.com/hyperhq/runv/blob/master/hypervisor/tty.go#L448
  *
- * \param infd File descriptor to read from
- * \param outfd File descriptor to write to
+ * \param shim \ref cc_shim
  */
 void
 handle_stdin(struct cc_shim *shim)
 {
-	ssize_t    nread;
-	char       buf[BUFSIZ-12];
-	char       wbuf[BUFSIZ];
-	int        ret;
-	ssize_t     len;
+	ssize_t      nread;
+	int          ret;
+	ssize_t      len;
+	static char  buf[BUFSIZ-STREAM_HEADER_SIZE] = {0};
+	static char  wbuf[BUFSIZ] = {0};
+	static bool  cr  = false;
+	static bool  emit = false;
+	static int   cnt = 0;
+	char         ch;
 
 	if (! shim || shim->proxy_io_fd < 0) {
 		return;
 	}
 
-	// write data to I/O fd
-	while ((nread = read(STDIN_FILENO, buf, BUFSIZ-12)) != -1) {
+	nread = read(STDIN_FILENO , &ch, 1);
```

Comment:
Is this single-byte `read()` purely to simplify the handling of special characters?